src/Agent.Listener/Configuration/ConfigurationManager.cs

```diff
@@ -805,5 +719,121 @@ private void CheckAgentRootDirectorySecure()
                 _term.Write(StringUtil.Loc("agentRootFolderCheckError"));
             }
         }
+
+        private bool SetupVstsProxySetting(IVstsAgentWebProxy vstsProxy, CommandSettings command) {
+            ArgUtil.NotNull(command, nameof(command));
+            
+            bool saveProxySetting = false;
+            string proxyUrl = command.GetProxyUrl();
+            if (!string.IsNullOrEmpty(proxyUrl))
+            {
+                if (!Uri.IsWellFormedUriString(proxyUrl, UriKind.Absolute))
+                {
+                    throw new ArgumentOutOfRangeException(nameof(proxyUrl));
+                }
+
+                Trace.Info("Reset proxy base on commandline args.");
+                string proxyUserName = command.GetProxyUserName();
+                string proxyPassword = command.GetProxyPassword();
+                (vstsProxy as VstsAgentWebProxy).SetupProxy(proxyUrl, proxyUserName, proxyPassword);
```

Comment:
In this realization result of casting `(vstsProxy as VstsAgentWebProxy)` can be null and it will throw an exception. It could be possible to use ?. to avoid it but as result, SetupProxy method may not be called. I think it will be better to inherit the new interface from IVstsAgentWebProxy with the needed methods and use it.

Reply:
Decided to add methods into interfaces